Threads.md

```diff
@@ -0,0 +1,471 @@
+| `i64.atomic.rmw16_s.cmpxchg` | 2 bytes | `expected` wrapped from i64 to i16, 16-bit compare equal | wrapped from i64 to i16, store 2 bytes | sign-extended from i16 to i64 |
+| `i64.atomic.rmw16_u.cmpxchg` | 2 bytes | `expected` wrapped from i64 to i16, 16-bit compare equal | wrapped from i64 to i16, store 2 bytes | zero-extended from i16 to i64 |
+| `i64.atomic.rmw32_s.cmpxchg` | 4 bytes | `expected` wrapped from i64 to i32, 32-bit compare equal | wrapped from i64 to i32, store 4 bytes | sign-extended from i32 to i64 |
+| `i64.atomic.rmw32_u.cmpxchg` | 4 bytes | `expected` wrapped from i64 to i32, 32-bit compare equal | wrapped from i64 to i32, store 4 bytes | zero-extended from i32 to i64 |
+| `i64.atomic.rmw.cmpxchg` | 8 bytes | 64-bit compare equal | 8 bytes | as i64 |
+
+### Thread operators
+
+  * `is_lock_free`: Given an operand `N` of type `i32`, if the atomic step of
+    an atomic primitive (see [Atomic Memory Accesses](#atomic-memory-accesses))
+    on a datum of size `N` bytes will be performed without the [agent][]
+    acquiring a lock outside the `N` bytes comprising the datum, then return
+    `1`. Otherwise, return `0`. Once the value of `is_lock_free` for a given
+    value `N` has been observed for any [agent][] in an [agent cluster][], it
+    cannot change.
+
+#### Wait and Wake operators
+
+The wake and wait operators are optimizations over busy-waiting for a value to
+change. It is a validation error to use these operators on non-shared linear
+memory. The operators have sequentially consistent ordering.
+
+Both wake and wait operators trap if the effective address of either operator
+is misaligned or out-of-bounds.
+
+The embedder is also permitted to suspend or wake an agent. A suspended agent
+can be woken by the embedder or the wake operator, regardless of how the agent
+was suspended (e.g. via the embedder or a wait operator).
+
+#### Wait
+
+The wait operator take three operands: an address operand, an expected value,
+and a relative timeout in milliseconds as an `f64`. The return value is `0`,
+`1`, or `2`, returned as an `i32`.
+
+| `timeout` value | Behavior |
+| ---- | ---- |
+| `timeout` <= 0 | Expires immediately |
+| 0 < `timeout` < Positive infinity | Expires after `timeout` milliseconds |
+| Positive infinity | Never expires |
+| NaN | Never expires |
+
+| Return value | Description |
+| ---- | ---- |
+| `0` | "ok", woken by another agent in the cluster or the embedder |
+| `1` | "not-equal", the loaded value did not match the expected value |
+| `2` | "timed-out", not woken before timeout expired |
+
+The wait operation begins by performing an atomic load from the given address.
+If the loaded value is not equal to the expected value, the operator returns 1
+("not-equal"). If the values are equal, the agent is suspended. If the agent
+is woken, the wait operator returns 0 ("ok"). If the timeout expires before
+another agent wakes this one, this operator returns 2 ("timed-out").
```

Comment:
I never really liked this [AgentCanSuspend](https://github.com/WebAssembly/design/pull/1019) compromise since it seems like it would just lead to spin-locking instead.  That being said, it seems like wasm should be symmetric with JS here, so either we make the restriction too or we try to remove the restriction from JS.  If we make the restriction, it seems like we'd throw a `RuntimeError`.